snuba/migrations/snuba_migrations/outcomes/0004_outcomes_matview_additions.py

```diff
@@ -0,0 +1,96 @@
+    Column("reason", String()),
+    Column("times_seen", UInt(64)),
+]
+
+new_materialized_view_columns: Sequence[Column[Modifiers]] = [
+    Column("org_id", UInt(64)),
+    Column("project_id", UInt(64)),
+    Column("key_id", UInt(64)),
+    Column("timestamp", DateTime()),
+    Column("outcome", UInt(8)),
+    Column("reason", String()),
+    Column("category", UInt(8)),
+    Column("quantity", UInt(64, Modifiers(nullable=True))),
+    Column("times_seen", UInt(64)),
+]
+
+
+class Migration(migration.MultiStepMigration):
+    """
+    Updates materialized view query to support category and quantity
+    """
+
+    blocking = False
+
+    def forwards_local(self) -> Sequence[operations.Operation]:
+        return [
+            operations.DropTable(
+                storage_set=StorageSetKey.OUTCOMES,
+                table_name="outcomes_mv_hourly_local",
+            ),
+            operations.CreateMaterializedView(
+                storage_set=StorageSetKey.OUTCOMES,
+                view_name="outcomes_mv_hourly_local",
+                destination_table_name="outcomes_hourly_local",
+                columns=new_materialized_view_columns,
+                query="""
+                    SELECT
+                        org_id,
+                        project_id,
+                        ifNull(key_id, 0) AS key_id,
+                        toStartOfHour(timestamp) AS timestamp,
+                        outcome,
+                        ifNull(reason, 'none') AS reason,
+                        category,
+                        count() AS times_seen,
+                        sum(quantity) AS quantity
+                    FROM outcomes_raw_local
+                    GROUP BY org_id, project_id, key_id, timestamp, outcome, reason, category
+                """,
+            ),
+        ]
+
+    def backwards_local(self) -> Sequence[operations.Operation]:
+        return [
+            operations.DropTable(
+                storage_set=StorageSetKey.OUTCOMES,
+                table_name="outcomes_mv_hourly_local",
+            ),
+            operations.CreateMaterializedView(
+                storage_set=StorageSetKey.OUTCOMES,
+                view_name="outcomes_mv_hourly_local",
+                destination_table_name="outcomes_hourly_local",
+                columns=old_materialized_view_columns,
+                query="""
+                    SELECT
+                        org_id,
+                        project_id,
+                        ifNull(key_id, 0) AS key_id,
+                        toStartOfHour(timestamp) AS timestamp,
+                        outcome,
+                        ifNull(reason, 'none') AS reason,
+                        count() AS times_seen
+                    FROM outcomes_raw_local
+                    GROUP BY org_id, project_id, key_id, timestamp, outcome, reason
+                """,
+            ),
+        ]
+
+    def forwards_dist(self) -> Sequence[operations.Operation]:
+        return []
```

Comment:
Doesn't the distributed table (outcomes_hourly_dist) also need to get updated with the new columns in the schema?

Reply:
I just saw this was added in migration 0003 - how come the column is being added to `outcomes_mv_hourly_local` after it is added to the distributed table, and in a separate migration. Is the state in between migrations 0003 and 0004 a valid one? (if not I would combine)

Reply:
The original migration was split up into two separate ones. the 0003 migration handles adding the additional columns to the raw and hourly table. 0004 here handles the matview changes. From what I see we don't set a material view on our distributed tables, so I assumed no changes were needed for the matview dist migrations. let me know if that makes sense!

Reply:
@lynnagara context here: https://sentry.slack.com/archives/C01KWV0PVDE/p1613521166009600?thread_ts=1613502457.007300&cid=C01KWV0PVDE

I will check that the state between the two migrations is valid and things still work as expected.

Reply:
What do you mean by "original migration was split up into two separate ones"? It looks like `0001_outcomes.py` creates all the tables in a single migration to me.

You are right about which columns need updating, ignore my first comment. But the order of execution is going to be: 0003 local migrations, 0003 dist migrations, then 0004 local migrations last (since its split over 2 migrations). I'm not sure if that ordering was intentional on your part.

Reply:
@lynnagara apologies on my ambiguity, by original migration i meant how i originally had it in the pull request.

Reply:
Right ok. So I guess my question is if you simply need to add what is essentially the same column in 3 spots and you're doing it all in one shot, why handle 2 in one migration, and the the last 1 in another migration. Is there something unique about this last case that I'm missing, or is the ordering that results from this (materialized after dist columns) somehow important?

Reply:
I'm not quite sure on the specifics of splitting up the changes into two migrations, @fpacifici could you elucidate? :)

Reply:
The second migration is a materialized view rebuild. It requires downtime no matter what. We have to stop the consumer in production. The first migraiton (adding the column) can be done without downtime.
There are chances we would be in a state in prod with the first migration fully applied but the second will have to wait.

Reply:
Regarding adding columns to the aggregated table (local and distributed) we need to add them before we make the change to the materialized view, and the state with the columns ready in the aggregated table but not filled is a valid one. We did the same for sessions.

Reply:
ah i did not realize the materialized view rebuild would require the consumers to be stopped. should i set `blocking = True` on the matview migration?

Reply:
Yes, please set blocking to True then. May be worth adding a note in the migration itself that the consumer needs to be stopped.